cli/tools/bench.rs

```diff
@@ -0,0 +1,541 @@
+#[serde(rename_all = "camelCase")]
+pub enum BenchResult {
+  Ok,
+  Ignored,
+  Failed(String),
+}
+
+#[derive(Debug, Clone, PartialEq, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct BenchPlan {
+  pub origin: String,
+  pub total: usize,
+  pub filtered_out: usize,
+  pub used_only: bool,
+}
+
+#[derive(Debug, Clone, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub enum BenchEvent {
+  Plan(BenchPlan),
+  Wait(BenchDescription),
+  Output(BenchOutput),
+  IterationStart(u64),
+  IterationFinish(u64),
+  Result(BenchDescription, BenchResult, u64),
+}
+
+#[derive(Debug, Clone)]
+pub struct BenchMeasures {
+  pub iterations: u64,
+  pub current_start: Instant,
+  pub measures: Vec<u128>,
+}
+
+#[derive(Debug, Clone)]
+pub struct BenchSummary {
+  pub total: usize,
+  pub passed: usize,
+  pub failed: usize,
+  pub ignored: usize,
+  pub filtered_out: usize,
+  pub measured: usize,
+  pub measures: Vec<BenchMeasures>,
+  pub current_bench: BenchMeasures,
+  pub failures: Vec<(BenchDescription, String)>,
+}
+
+impl BenchSummary {
+  pub fn new() -> Self {
+    Self {
+      total: 0,
+      passed: 0,
+      failed: 0,
+      ignored: 0,
+      filtered_out: 0,
+      measured: 0,
+      measures: Vec::new(),
+      current_bench: BenchMeasures {
+        iterations: 0,
+        current_start: Instant::now(),
+        measures: vec![],
+      },
+      failures: Vec::new(),
+    }
+  }
+
+  fn has_failed(&self) -> bool {
+    self.failed > 0 || !self.failures.is_empty()
+  }
+
+  fn has_pending(&self) -> bool {
+    self.total - self.passed - self.failed - self.ignored > 0
+  }
+}
+
+pub trait BenchReporter {
+  fn report_plan(&mut self, plan: &BenchPlan);
+  fn report_wait(&mut self, description: &BenchDescription);
+  fn report_output(&mut self, output: &BenchOutput);
+  fn report_result(
+    &mut self,
+    description: &BenchDescription,
+    result: &BenchResult,
+    elapsed: u64,
+    current_bench: &BenchMeasures,
+  );
+  fn report_summary(&mut self, summary: &BenchSummary, elapsed: &Duration);
+}
+
+struct PrettyBenchReporter {
+  echo_output: bool,
+}
+
+impl PrettyBenchReporter {
+  fn new(echo_output: bool) -> Self {
+    Self { echo_output }
+  }
+
+  fn force_report_wait(&mut self, description: &BenchDescription) {
+    print!(
+      "bench {} ... {} iterations ",
+      description.name, description.iterations
+    );
+    // flush for faster feedback when line buffered
+    std::io::stdout().flush().unwrap();
+  }
+}
+
+/// A function that converts a milisecond elapsed time to a string that
+/// represents a human readable version of that time.
+fn human_elapsed(elapsed: u128) -> String {
+  if elapsed < 1_000 {
+    return format!("({}ms)", elapsed);
+  }
+  if elapsed < 1_000 * 60 {
+    return format!("({}s)", elapsed / 1000);
+  }
+
+  let seconds = elapsed / 1_000;
+  let minutes = seconds / 60;
+  let seconds_remainder = seconds % 60;
+  format!("({}m{}s)", minutes, seconds_remainder)
+}
```

Comment:
We should share this with `test.rs` since this function is already in there and tested there.

Reply:
Any suggestion where do we put it? I also have `human_size`, maybe let's add `cli/display.rs` for such utilities?

Reply:
Opened https://github.com/denoland/deno/pull/13879